membership_extension/models/account_invoice.py

```diff
@@ -18,23 +19,42 @@ def action_cancel_draft(self):
 
     @api.multi
     def action_cancel(self):
-        self.mapped('invoice_line').mapped('membership_lines').write({
+        """Cancel membership for customer invoices and restore previous
+        membership state for customer refunds. Harmless on supplier ones.
+        We detect dynamically if the module account_refund_original is
+        installed for accurate source invoice.
+        """
+        # TODO: Module in v9 is `account_invoice_refund_link`
+        self.mapped('invoice_line.membership_lines').write({
             'state': 'canceled',
         })
-        for refund in self.filtered(lambda r: r.type == 'out_refund'):
-            origin = self.search([
-                ('type', '=', 'out_invoice'),
-                ('number', '=', refund.origin),
-            ])
-            lines = origin.mapped('invoice_line').mapped('membership_lines')
-            if origin and lines:
-                origin_state = 'paid' if origin.state == 'paid' else 'invoiced'
-                lines.filtered(lambda r: r.state == 'canceled').write({
-                    'state': origin_state,
-                })
-                lines.write({
-                    'date_cancel': False,
-                })
+        refunds = self.filtered(lambda r: (
+            r.type == 'out_refund' and
+            # TODO: Rename this to `origin_invoice_ids` on v9
+            (r.origin or getattr(r, 'origin_invoices_ids', False))
```

Comment:
As a rule of thumb, you should avoid using `getattr` to access fields, [as you can see in slide 41 of the safe coding tips presentation given in last Odoo Experience](https://www.odoo.com/es_ES/slides/slide/10-rules-for-safer-code-398).

Normally it should be `'origin_invoices_ids' in r and r.origin_invoices_ids`.

However, it does not seem a blocker, since the field name is hardcoded anyway 😉